data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,562 @@
+```ocaml
+# type 'a tree = Node of 'a * 'a tree list;;
+type 'a tree = Node of 'a * 'a tree list
+
+# let rec tree_map f (Node (x, u)) = Node (f x, List.map (tree_map f) u);;
+val tree_map : ('a -> 'b) -> 'a tree -> 'b tree = <fun>
+```
+
+Because records are implicitly single constructor variants, this also applies to them:
+```ocaml
+# type name = { first : string; last: string };;
+type name = { first : string; last : string; }
+
+# let robin = { first = "Robin"; last = "Milner" };;
+val robin : name = {first = "Robin"; last = "Milner"}
+
+# let { first; last } = robin;;
+val first : string = "Robin"
+val last : string = "Milner"
+```
+
+A special case of combined definition and pattern matching involves the unit type:
+```ocaml
+# let () = print_endline "ha ha";;
+ha ha
+```
+
+Below, the pattern does not contain any identifier, meaning no name is defined. The expression is evaluated, the side effect takes place, no definition is created, and no value is returned. Writing that kind of pseudo-definition only expresses interest in the side effects.
+```ocaml
+# print_endline "ha ha";;
+ha ha
+- : unit = ()
+
+# let _ = print_endline  "ha ha";;
+ha ha
+- : unit = ()
+```
+
+As seen in the last example, the catch-all pattern (the underscore symbol) can be used in definitions. The following example illustrates its use, which is distinct from the unit pattern:
+```ocaml
+# let (_, y) = List.split [(1,2); (3,4); (5,6); (7,8)];;
+val y : int list = [2; 4; 6; 8]
+```
+
+### Local Definitions
+
+Local definitions are like global definitions, except the name is only bound inside an expression. They are introduced by the `let ... = ... in ...` construction. The name bound before the `in` keyword is only bound in the expression after the `in` keyword
+```ocaml
+# let b = 2 * 3 in b * 7;;
+- : int = 42
+
+# b;;
+Error: Unbound value b
+```
+
+Here, the name `b` is bound to `6` inside the expression `b * 7`. A couple of remarks:
+- No global definition is introduced in this example
+- Computation of `2 * 3` will always take place before `b * 7`
+
+Local definitions can be chained (one after another) or nested (one inside another). Here is an example of chaining:
+```ocaml
+# let b = 2 * 3 in
+  let c = b * 7 in
+  b * c;;
+- : int = 252
+```
+
+This is how scoping works here:
+- `b` is bound to `6` inside `let c = b * 7 in b * c`
+- `c` is bound to `42` inside `b * c`
+
+Here is an example of nesting:
+```ocaml
+# let b =
+    let c = 2 * 3 in
+    c * 5 in
+  b * 7;;
+- : int = 210
+Here is how scoping works:
+- `c` is bound to `6` inside `c * 5`
+- `b` is bound to `30` inside `b * 7`
+
+Arbitrary combinations or chaining and nesting are allowed.
+
+### Scopes and Environments
+
+Without oversimplifying, an OCaml program is a sequence of definitions or expressions. These items are said to be at the _toplevel_. That's why an OCaml REPL, such as UTop, is called a toplevel, because that's where typed expressions go.
```

Comment:
```suggestion
Without oversimplifying, an OCaml program is a sequence of definitions or expressions. These items are said to be at the _top level_. That's why an OCaml REPL, such as UTop, is called a toplevel, because that's where typed expressions go.
```

Perhaps separate "top level," where they appear, and "toplevel," the REPL.